New NeuroEvolution Functionality

shiffman opened the pull request:
🌈DEAR BELOVED ML5 COMMUNITY, 🌈

I'm making a Pull Request(PR). Please see the details below.

### → Which branch are you submitting to? 🌲

This is a work-in-progress branch which should go into `development` before release.

### → Describe your Pull Request 📝

It's a brand new feature for `ml5.neuralNetwork()` allowing for "neuro-evolution" examples as previously discussed in #113! I have previously made tutorials and examples around neuroevolution, most recently with TensorFlow.js without ml5.

[![cdUNkwXx-I4hd](https://user-images.githubusercontent.com/191758/78712760-1dbe2300-78e7-11ea-9d83-364e4b2031f0.jpg)](https://www.youtube.com/watch?v=cdUNkwXx-I4)

This pull request is the start of re-creating this example but instead of manually building a model with tf.js I am repurposing `m5.neuralNetwork()`. 

### → Example 🦄

A working example can be found in `examples/p5js/NeuralNetwork/NeuroEvolution`.

![neuro](https://user-images.githubusercontent.com/191758/78708083-8c977e00-78df-11ea-9f7e-ba856e455524.gif)

### → Questions ❓

This is in progress and there are few items that I need to address. I will work on these and post updates in this thread, but open to any and all ideas / feedback.

#### Creating a Neural Network with no data

The neuroevolution process does not involve training data. In fact, the neural network is never trained. A population of neural networks is generated with random weights and these weights are "evolved." I would like to be able to write code as follows:

```javascript
const options = {
  inputs: 3,
  outputs: ['yes', 'no'],
  task: 'classification'
}
const nn = ml5.neuralNetwork(options);
nn.classify([0.2, 0.1, 0.9], (err, result) => {
  console.log(result);
});  
```
This does not work with `0.5.0` as `ml5.neuralNetwork()` requires data for the model to be built. The data then forms the basis for configuring the `meta` object in `NeuralNetworkData`. Without it, the neural network layers and associated weights are not generated.

My test example has a workaround right now but I'll need to figure out how to force `ml5.NeuralNetwork` to build the model without data. Another function or perhaps a flag in `options`?  Maybe if the task is `"neuroevolution"`? Something like. . ?

```javascript
const options = {
  inputs: 3,
  outputs: ['yes', 'no'],
  task: 'classification',
  data: 'none'
}
const nn = ml5.neuralNetwork(options);
```

or `task: 'neuroevolution'`?

#### Synchronous Prediction / Classification

My examples involve calling `predict()` many times (for each member of the population) each cycle through `draw()`. This is very difficult to manage and stay in sync with the animation loop using callbacks. I have added a `predictSync()` and `classifySync()` that makes use of tf.js `arraySync()`. Another way I could handle this is with batch processing all of the population using `predictMultiple()` and waiting for the callback to move onto the next animation step. 

```javascript
const nn = ml5.neuralNetwork(options);
let results = nn.classifySync([0.2, 0.1, 0.9]);
```

I am not sure how best to proceed here and will continue to experiment.

Related to this I am also forcing tf.js to use `cpu` as it performs really well with these synchronous calls to `predict()` with little bits of data. Maybe an `ml5.util` method to force cpu backend would be helpful, or perhaps it exists already!

More to come!

shiffman commented:
Worked on this a bit more today during a Coding Train members stream. Here is the status:

### mutate() works!

You can now `mutate()` a Neural Network, by calling:

```javascript
  const nn = ml5.neuralNetwork(options);
  nn.neuralNetwork.mutate(0.1);
```

The argument is the rate of mutation (10% above). Next steps for this are:

* [ ] this should be called on `nn` not `nn.neuralNetwork()`
* [ ] need a better default way of mutating the weights (Gaussian distribution?)

### Creating model w/o Data still unresolved

The recent commit includes a new function aptly named `buildModelBasedOnNothing()`. This is a temporary solution that creates dummy data based on the `options` presented to `ml5.neuralNetwork()`. 

```javascript
  const options = {
    inputs: 3,
    outputs: ['yes', 'no'],
    task: 'classification'
  }
  const nn = ml5.neuralNetwork(options);
  nn.buildModelBasedOnNothing();
```

Another option would be to have the end-user offer the "dummy data" themselves? e.g.:

```javascript
  const options = {
    task: 'classification'
  }
  const nn = ml5.neuralNetwork(options);
  nn.addSampleData([0,0,0], ['yes']);
  nn.addSampleData([0,0,0], ['no']);
  nn.finishSampleData();
```

or sample data in options?

```javascript
  const options = {
    task: 'classification',
    sampleData: [[[0,0,0], ['yes']], [[0,0,0], ['no']]]
  }
  const nn = ml5.neuralNetwork(options);
```

On purpose I'm naming everything above terribly b/c I don't have a good idea for names yet.

shiffman commented:
Hey! This is much further along and almost ready for a full review / merge. Quick question first though @joeyklee maybe you have a thought? Currently, the way that the example works everytime I create a new population of neural networks, I [manually dispose the memory for the old population](https://github.com/ml5js/ml5-library/blob/neuroevolution/examples/p5js/NeuralNetwork/NeuroEvolution/bird.js#L31) in the example code itself.

```javascript
dispose() {
  this.brain.neuralNetwork.model.dispose();
}
```

I suppose this has not come up before for "regular" ml5 examples b/c there is typically just one neural network that lives for the duration of the web application. I could add a `dispose()` method to `ml5.neuralNework()` or might there be a more elegant way for ml5 to handle this behind the scenes?

joeyklee commented:
> Hey! This is much further along and almost ready for a full review / merge. Quick question first though @joeyklee maybe you have a thought? Currently, the way that the example works everytime I create a new population of neural networks, I [manually dispose the memory for the old population](https://github.com/ml5js/ml5-library/blob/neuroevolution/examples/p5js/NeuralNetwork/NeuroEvolution/bird.js#L31) in the example code itself.
> 
> ```js
> dispose() {
>   this.brain.neuralNetwork.model.dispose();
> }
> ```
> 
> I suppose this has not come up before for "regular" ml5 examples b/c there is typically just one neural network that lives for the duration of the web application. I could add a `dispose()` method to `ml5.neuralNework()` or might there be a more elegant way for ml5 to handle this behind the scenes?

+ For now do you think it would make sense to add method directly to the class that will handling this so that people can do something like: `nn.resetNetwork()`? or something along those lines?

+ For future reference: I was thinking maybe we can add to the `neuralNetwork` class an array of "brains" that store all the blank networks?

shiffman commented:
Thanks @joeyklee for the reply!

> For now do you think it would make sense to add method directly to the class that will handling this so that people can do something like: nn.resetNetwork()? or something along those lines?

Yes, although I think `resetNetwork()` would be something different, that to me indicates just randomizing all the weights like a brand new network. What I'm talking about is actually disposing the memory and releasing the tensors. I think I'll just add a `dispose()` method to the neural network class for now!

> For future reference: I was thinking maybe we can add to the neuralNetwork class an array of "brains" that store all the blank networks?

Yes, this is exactly what I am thinking too! I was considering something like this "Phase 2" of my neuroevolution exploration and would likely be a new class rather than part of `ml5.neuralNetwork()`? I would work on this as a follow-up to this pull request. 

It might be something like `ml5.ga()` (for "genetic algorithm") or `ml5.population()` or `ml5.neuroEvolution()`. This new class would manage a population (e.g. array) of neural networks and include functions selection and reproduction.

shiffman commented:
Ok, this pull request is ready for review! I would say the main unresolved aspect here for me is the addition of `classifySync()` and `predictSync()`.  I just duplicated `predict()` and `classify()` and changed one line to make them synchronous. I'm assuming there is a better way to refactor these? I'm happy to take a crack at it, but curious for any thoughts you have @bomanimc in terms of how you have been managing the code base!

Any other feedback welcome and appreciated!

I will be teaching this material later this week in my ITP course and will use a custom build of the library. Other bugs and questions will likely come up then!

joeyklee commented:
Wow! This is exciting. I will take a deeper look tomorrow and test on my side. I might suggest merging for now if your examples run and the other NN examples are still working correctly. 

In seeing this PR, I'm reminded that I think it would make sense to do a 3rd pass at modularizing and cleaning up the NN class. It might also make sense to enforce some kind of MVC style here. I'd be happy to explore this!

joeyklee commented:
> Yes, this is exactly what I am thinking too! I was considering something like this "Phase 2" of my neuroevolution exploration and would likely be a new class rather than part of ml5.neuralNetwork()? I would work on this as a follow-up to this pull request.
> 
> It might be something like ml5.ga() (for "genetic algorithm") or ml5.population() or ml5.neuroEvolution(). This new class would manage a population (e.g. array) of neural networks and include functions selection and reproduction.

+ I think this would make sense. We could call the parts of the ml5.neuralNetwork() into other features that rely on the same functions. 
+ this makes me think that would actually separate out a lot of the NeuralNetwork sub classes (e.g. NeuralNetworkData) into more generic components that can be accessed elsewhere -- basically build an API that makes things more clean and consistent.

joeyklee commented:
> Ok, this pull request is ready for review! I would say the main unresolved aspect here for me is the addition of classifySync() and predictSync(). I just duplicated predict() and classify() and changed one line to make them synchronous. I'm assuming there is a better way to refactor these? I'm happy to take a crack at it, but curious for any thoughts you have @bomanimc in terms of how you have been managing the code base!

+ I think for now this is fine. I would go for merging this in with the idea that we will do a review and refactor for another PR.

joeyklee commented:
@shiffman - Before we merge, just a few notes:
1. Please append a description to the example `NeuroEvolution` => `NeuroEvolution_FlappyBird` ?
2. For future reference, maybe it makes sense to have the flappy birth example as a demo rather than an example since it is more complex than a raw example? 

Merge when ready.

shiffman commented:
Ok, I updated the flappy bird example to `NeuroEvolution_FlappyBird` as suggested, cleaned up the code a bit, and commented key sections. I agree @joeyklee this one likely falls under the "demo" category! It's hard to come up with a nice really simple neuroevolution example! I have a couple ideas that I will explore in a separate pull request.

If this all looks good, we can merge!

joeyklee commented:
Thanks @shiffman - Feel free to merge when ready.

joeyklee commented:
@shiffman - I will go ahead and merge. The tests are timing out and need to be fixed! We can follow up on this later.

ozzel commented:
I am looking for a solution/approach to track a specific object in video, for example guitar while musician is playing, to get object coordinates just like pose detection, but for custom object tracking.  How should I start, do you have and advice for any documentation using p5 and ml5?

LukeTheWalker commented:
I would have chosen to have the crossover function out of the NeuralNetwork class, as it makes more sense since its result is independent from which NN you call the function from.

Also, may consider adding a more explanatory definition of the 'other' parameter in the crossover function, something like 'the other NeuralNetwork to crossover with'

shiffman commented:
thank you @LukeTheWalker! These are excellent suggestions! Can you file this as an issue perhaps? Comments here in a merge pull request may get lost as it's not tracked as an open item.
